pkg/network/netlink/attribute_scanner.go

```diff
@@ -0,0 +1,206 @@
+	frames []*NestedFrame
+}
+
+// A NestedFrame encapsulates the decoding information of a certain nesting level
+type NestedFrame struct {
+	// The current attribute we're looking at
+	attr netlink.Attribute
+
+	// The slice of input bytes and its iterator index.
+	b []byte
+	i int
+
+	// Any error encountered while decoding attributes.
+	err error
+}
+
+// NewAttributeScanner returns a new instance of AttributeScanner
+func NewAttributeScanner() *AttributeScanner {
+	// We pre-allocate 3 frames since the Conntrack messages have 3 levels of nesting
+	scanner := &AttributeScanner{
+		frames: make([]*NestedFrame, 3),
+	}
+
+	for i := range scanner.frames {
+		scanner.frames[i] = &NestedFrame{}
+	}
+
+	return scanner
+}
+
+// Next advances the scanner to the next netlink attribute (within the same NestedFrame).
+// It returns false when no more attributes are present, or an error was encountered.
+func (s *AttributeScanner) Next() bool {
+	f := s.frame()
+
+	if f.err != nil {
+		// Hit an error, stop iteration.
+		return false
+	}
+
+	// Exit if array pointer is at or beyond the end of the slice.
+	if f.i >= len(f.b) {
+		return false
+	}
+
+	if err := f.unmarshal(); err != nil {
+		f.err = err
+		return false
+	}
+
+	// Advance the pointer by at least one header's length.
+	if int(f.attr.Length) < nlaHeaderLen {
+		f.i += nlaHeaderLen
+	} else {
+		f.i += nlaAlign(int(f.attr.Length))
+	}
+
+	return true
+}
+
+// Type returns the Attribute.Type field of the current netlink attribute
+// pointed to by the scanner.
+func (s *AttributeScanner) Type() uint16 {
+	// Mask off any flags stored in the high bits.
+	return s.frame().attr.Type & attrTypeMask
+}
+
+// Err returns the first error encountered by the scanner.
+func (s *AttributeScanner) Err() error {
+	return s.frame().err
+}
+
+// Bytes returns the raw bytes of the current Attribute's data.
+func (s *AttributeScanner) Bytes() []byte {
+	return s.frame().attr.Data
+}
+
+// Nested executes the given function within a new NestedFrame
+func (s *AttributeScanner) Nested(fn func() error) {
+	// Push new frame
+	s.push()
+
+	// Execute function within new frame
+	err := fn()
+
+	// Pop frame and assign error
+	s.pop()
+	s.frame().err = err
+}
+
+// ResetTo makes the current AttributeScanner ready for another netlink message
+func (s *AttributeScanner) ResetTo(data []byte) error {
+	s.level = 0
+
+	for _, f := range s.frames {
+		f.b = nil
+		f.i = 0
+		f.err = nil
+	}
+
+	if len(data) < 2 {
+		return errMessageTooShort
+	}
+
+	offset := messageOffset(data[:2])
+
+	if len(data) <= offset {
+		return errMessageTooShort
+	}
+
+	s.frame().b = data[offset:]
+	return nil
+}
+
+// frame returns the current frame
+func (s *AttributeScanner) frame() *NestedFrame {
+	return s.frames[s.level]
+}
+
+// push a new nested frame
+func (s *AttributeScanner) push() {
+	prev := s.frame()
+
+	// Create a new frame object if necessary
+	if len(s.frames) <= s.level {
+		s.frames = append(s.frames, &NestedFrame{})
+	}
```

Comment:
I just wanted for the attribute scanner to work with any type of message. This ensures that it won't panic if used with more level's of nesting. Maybe I should just remove the pre-allocation in the constructor instead, to avoid any confusion.